source/log.c

```diff
@@ -0,0 +1,240 @@
+*/
+
+#include <stdio.h>
+
+#include <aws/common/atomic.h>
+#include <aws/common/common.h>
+#include <aws/common/log.h>
+#include <aws/common/memory_pool.h>
+#include <aws/common/mutex.h>
+#include <aws/common/thread.h>
+
+static void aws_log_default_report_function_internal(const char* log_message) {
+    (void)log_message;
+}
+
+int global_log_thread_count;
+aws_log_report_callback global_log_report_callback = aws_log_default_report_function_internal;
+struct aws_mutex global_log_list_mutex = AWS_MUTEX_INIT;
+struct aws_log_context *global_log_list;
+
+AWS_THREAD_LOCAL struct aws_log_context *thread_local_log_context;
+
+void aws_log_set_reporting_callback(aws_log_report_callback report_callback) {
+    if (report_callback) {
+        global_log_report_callback = report_callback;
+    } else {
+        global_log_report_callback = aws_log_default_report_function_internal;
+    }
+}
+
+int aws_vlog(enum aws_log_level level, const char *fmt, va_list va_args) {
+    (void)level;
+
+    if (!thread_local_log_context) {
+        aws_raise_error(AWS_ERROR_LOG_UNINITIALIZED);
+        return AWS_OP_ERR;
+    }
+
+    if (!thread_local_log_context->running) {
+        aws_raise_error(AWS_ERROR_LOG_UNINITIALIZED);
+        return AWS_OP_ERR;
+    }
+
+    /* Atomically grab and release all messages on the `delete_list` whenever present. */
+    struct aws_log_message* delete_list;
+    do {
+        delete_list = thread_local_log_context->delete_list;
+    } while (!aws_atomic_compare_exchange_ptr((void **)&thread_local_log_context->delete_list, delete_list, NULL));
+
+    while (delete_list) {
+        struct aws_log_message *next = delete_list->next;
+        aws_memory_pool_release(&thread_local_log_context->message_pool, delete_list);
+        delete_list = next;
+    }
+
+    /* Acquire memory for the new message. */
+    void *mem = aws_memory_pool_acquire(&thread_local_log_context->message_pool);
+    struct aws_log_message *msg = (struct aws_log_message *)mem;
+    char *msg_data = (char*)(msg + 1);
+
+    /* Format the message. */
+    vsnprintf((char *)msg_data, thread_local_log_context->max_message_len, fmt, va_args);
+
+    /* Push formatted message onto this thread's message list. Will be picked up later and
+    processed by the logging thread in `aws_log_flush()`, and handed to a user callback set by
+    `aws_log_set_reporting_callback`. */
+    do {
+        msg->next = thread_local_log_context->message_list;
+    } while (!aws_atomic_compare_exchange_ptr((void **)&thread_local_log_context->message_list, msg->next, msg));
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_log(enum aws_log_level level, const char *fmt, ...) {
+    va_list va_args;
+    va_start(va_args, fmt);
+    int ret = aws_vlog(level, fmt, va_args);
+    va_end(va_args);
+    return ret;
+}
+
+const char *aws_log_level_to_string(enum aws_log_level level) {
+    switch (level) {
+        case AWS_LOG_LEVEL_OFF:     return "AWS_LOG_LEVEL_OFF";

```

Comment:
depending on use case, these likely need to be just "OFF" , "FATAL" etc...

I say that hoping that these strings are part of the actual logged message.

Reply:
They are not apart of the actual message yet. I can add this functionality in.